src/lib/optimizer/optimizer.cpp

```diff
@@ -172,11 +174,31 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(std::shared_ptr<AbstractLQP
   const auto root_node = LogicalPlanRootNode::make(std::move(input));
   input = nullptr;
 
-  if constexpr (HYRISE_DEBUG) validate_lqp(root_node);
+  if constexpr (HYRISE_DEBUG) {
+    validate_lqp(root_node);
 
-  for (const auto& rule : _rules) {
-    _apply_rule(*rule, root_node);
-    if constexpr (HYRISE_DEBUG) validate_lqp(root_node);
+    Timer rule_timer{};
+    for (const auto& rule : _rules) {
+      _apply_rule(*rule, root_node);
+      auto rule_duration = rule_timer.lap();
+
+      auto& r = *rule.get();
+      auto rule_name = std::string(typeid(r).name());
+      rule_name = rule_name.substr(
+          rule_name.rend() - std::find_if(rule_name.rbegin(), rule_name.rend(), [](auto c) { return std::isdigit(c); }),
+          rule_name.size());
+      rule_name = rule_name.substr(0, rule_name.find("Rule") + 4);
+
+      if (rule_durations) {
+        rule_durations->push_back({rule_name, rule_duration});
```

Comment:
This looks like it's not done for release builds?

Reply:
indeed. that's why we check for HYRISE_DEBUG earlier